README.md
<br/> <p style="color:green">chunkR</p> devel <img src  = "https://github.com/leandroroser/chunkR/blob/master/inst/extdata/blank.png" height = "100px" width="100px" align="right"  alt = "blank"/>  <img src  = "https://github.com/leandroroser/chunkR/blob/master/inst/extdata/icon.png" width="230px" align="right"  alt = "chunkR icon"/> 
======================================================


[![Build Status](https://travis-ci.org/leandroroser/chunkR.svg?branch=master)](https://travis-ci.org/leandroroser/chunkR) [![AppVeyor Build Status](https://ci.appveyor.com/api/projects/status/github/leandroroser/chunkR?branch=master&svg=true)](https://ci.appveyor.com/project/leandroroser/chunkR) [![R](https://www.r-pkg.org/badges/version/chunkR)](https://CRAN.R-project.org/package=chunkR) [![Coverage status](https://codecov.io/gh/leandroroser/chunkR/branch/master/graph/badge.svg)](https://codecov.io/github/leandroroser/chunkR?branch=master)


<br/><br/><br/><br/><br/><br/>
This package allows to read large text tables in chunks, using a fast C++ backend. Text files can be imported as data frames (with automatic column type detection option) or matrices. The program is designed to be simple and user-friendly. 


chunkR is based on three basic functions: <span style="color:green; weight:bold">chunker</span>, to create a connection to a text file; <span style="color:green; weight:bold">next_chunk()</span> to read the next chunk, and <span style="color:green; weight:bold">get_table()</span> to retrieve the corresponding data chunk.


These functions can be easily included in loops and other source code, using the return value of the next_chunk() function, that is TRUE when a new chunk is available and FALSE when the file was totally read, respectively. The get_table() function, returns an empty data frame/matrix when next_chunk() is FALSE. See the examples below.


<span style="color:blue">Installation<span style="color:blue">
------------

```R
library(devtools)
install_github("leandroroser/chunkR")
```

<span style="color:blue">Examples<span style="color:blue">
--------

```R
data(iris)

# write iris as tab delimited file. Note that quote is set to FALSE
tmp_path <- file.path(tempdir(),"iris.txt")
write.table(iris, tmp_path, quote = FALSE)


#-----------------------------------------------------------------#
#--- Reading a data frame with automatic column-type detection ---#
#-----------------------------------------------------------------#


# create a 'chunker' object passing the path of the input file.
my_chunker_object <- chunker(tmp_path, chunksize = 30)

# read a chunk
next_chunk(my_chunker_object)

# get the chunk
 get_table(my_chunker_object)

# read another chunk
next_chunk(my_chunker_object)

# get the number of lines already read
get_completed(my_chunker_object)


#---- Quoted data --------#


write.table(iris, tmp_path, quote = TRUE)

my_chunker_object <- chunker(tmp_path, quoted = TRUE, chunksize = 30)

next_chunk(my_chunker_object)

get_table(my_chunker_object)


#---- Data without rownames and/or colnames ----#


tmp_path <- file.path(tempdir(),"iris.txt")
write.table(iris, tmp_path, row.names = FALSE, col.names = FALSE)

my_chunker_object2 <- chunker(tmp_path, quoted = TRUE, chunksize = 30,
has_rownames = FALSE, has_colnames = FALSE)

next_chunk(my_chunker_object2)

get_table(my_chunker_object2) # automatic generation of rownames and/or colnames


#--- read a csv file ---#


tmp_path_csv <- file.path(tempdir(),"iris.csv")

write.table(iris, tmp_path_csv, quote = FALSE, sep = ",")

# read the csv indicating the value of the sep parameter
my_chunker_object3 <- chunker(tmp_path_csv, chunksize = 30, sep = ",")
# the file can  then be processed as with tab delimiters

next_chunk(my_chunker_object3)
get_table(my_chunker_object3)

# remove temporal file
file.remove(tmp_path_csv)


#--------------------------------------------------------#
#--- Reading a data frame using column types argument ---#
#--------------------------------------------------------#


## Four types can be passed : "character", "numeric" (aka "double"), "integer", "logical"

# create a 'chunker' object passing the path of the input file.
my_chunker_object4 <- chunker(tmp_path, chunksize = 120,
 columns_classes = c("numeric", "numeric", "numeric","numeric", "character"))

# read a chunk
next_chunk(my_chunker_object4)

# get the chunk
get_table(my_chunker_object4)

# read another chunk
next_chunk(my_chunker_object4)

# get the number of lines already read
get_completed(my_chunker_object4)


#-------------------------#
#--- Reading a matrix  ---#
#-------------------------#


my_chunker_object5 <- chunker(tmp_path, chunksize = 30, data_format= "matrix")

# read a chunk
next_chunk(my_chunker_object5)

# store the chunk as a character matrix in R
this_data <- get_table(my_chunker_object5)


# The package provides a fast generic C++ function for conversion from
# matrix (any R type) to data frame
this_data_as_df2 <- matrix2df(this_data)

# remove temporal file
file.remove(tmp_path)


#----------------------------------#
#--- Example with a big table -----#
#----------------------------------#


### Example with a data frame

# create a large data frame, and write it in a temporal directory

tmp_path <- file.path(tempdir(),"big_table.txt")

out <- data.frame(numeric_data = runif(1000000),
                  character_data = sample(c("a", "t", "c", "g"), 1000000, 
                  replace = TRUE),
                  integer_data = sample(1000000),
                  bool_data = sample(c(TRUE, FALSE), 1000000, replace = TRUE))


write.table(out, tmp_path, quote = FALSE)

# create a chunker object, reading in chunks of 10000 lines
my_chunker_object6 <- chunker(tmp_path, chunksize = 10000)

next_chunk(my_chunker_object6)
data <- get_table(my_chunker_object6) 

# check classes
lapply(data,typeof)
file.remove(tmp_path)


### Example with a matrix

# create a large matrix, and write it in a temporal directory

my_table <- tempfile()
write.table(matrix(sample(c("a", "t", "c", "g"), 1000000, replace = TRUE), 
100000, 1000), my_table, quote = FALSE)

# create a chunker object, reading in chunks of 10000 lines
my_chunker_object7 <- chunker(my_table, chunksize = 10000, data_format= "matrix")

# create a loop to read all the file and do something with it

lines <- 0
while(next_chunk(my_chunker_object7))
{
  data <- get_table(my_chunker_object7) 
  
  # do something with data, e.g., convert to data frame first
  data <- matrix2df(data)
  
  lines <- lines + nrow(data)
  cat("Processed ", lines, "lines\n")
}

# remove the temporal file
file.remove(my_table)


```


NEWS
chunkR 1.1.1 (devel)
==============
Changes:
* Added detection of end of line character for cross platform compatibility.
* Correction of return values in exported functions.

chunkR 1.1.0 (Release date: 2018-01-21)
==============
Changes:
* The reader class was renamed as chunker for consistency with the package.
* The reader function was renamed as chunker.
* New functionality to read dataframes direcly from the source file.
  The newinterface does not requires to read the data as a text matrix 
  prior to the obtention of a dataframe.
* The functions get_matrix and get_dataframe where converted into internals
  and substituted at user level by the funcion get_table that can retrieve
  both dataframe and matrix data types. 
* Function get_matrix2dataframe was deprecated.
* Correction of bugs.

R/constructors.R

#' chunker
#' @param path Input file path 
#' @param sep Character separating cells in the input table (default = " ")
#' @param quoted Quoted character data? Default FALSE. If TRUE, the program removes quotes.
#' @param has_colnames Column names present in the input table? (Logical, default TRUE)
#' @param has_rownames Row names present in the input table? (Logical, default TRUE)
#' @param chunksize Chunk size (default 1000)
#' @param data_format Format of input data: "data.frame" (default) or "matrix".
#' @param columns_classes Vector with columns classes: "character", "numeric" (aka "double"), 
#' "integer" or "logical".
#' @param autodetect Use auto-detection of columns classes? Default TRUE.
#' @param scan_rows How many rows to scan for auto-detection of columns classes. Default 10. Note
#' that this value shoud be increased when columns present all NA values in the first scan_rows ranger of rows.
#' Checking is perfomed via a call to read.table with the scan_rows value passed to the nrows parameter,
#' and the posterior extraction of the corresponding classes detected in this call. 
#' 
#' @description The objects of class "chunker" are the central elements of the
#' chunkR package. These objects can store a data chunk and other 
#' information required for the process of reading a file in pieces.
#' A chunker object is created with the chunker() function, 
#' that requires the path to a file, and other arguments, as the size of the chunk 
#' and the data type ("data.frame" or "matrix").
#' Two basic methods are defined to manipulate the object: 
#' 
#' -  \code{\link{next_chunk}} function to read the next chunk
#' 
#' -  \code{\link{get_table}} function to retrieve the data
#' 
#' The functions \code{\link{get_completed}} and \code{\link{get_colnames}}
#' allow to get the number of rows already read, and the column names of the 
#' table.
#' 
#' @examples
#' 
#' 
#' data(iris)
#' 
#' # write iris as tab delimited file. Note that quote is set to FALSE
#' 
#' tmp_path <- file.path(tempdir(),"iris.txt")
#' write.table(iris, tmp_path, quote = FALSE)
#' 
#' #-----------------------------------------------------------------#
#' #--- Reading a data frame with automatic column-type detection ---#
#' #-----------------------------------------------------------------#
#' 
#' # create a 'chunker' object passing the path of the input file.
#' my_chunker_object <- chunker(tmp_path, chunksize = 30)
#' 
#' # read a chunk
#' next_chunk(my_chunker_object)
#' 
#' # get the chunk
#'  get_table(my_chunker_object)
#' 
#' # read another chunk
#' next_chunk(my_chunker_object)
#' 
#' # get the number of lines already read
#' get_completed(my_chunker_object)
#' 
#' 
#' #--- read a csv file ---#
#' 
#' tmp_path_csv <- file.path(tempdir(),"iris.csv")
#' 
#' write.table(iris, tmp_path_csv, quote = FALSE, sep = ",")
#' 
#' # read the csv indicating the value of the sep parameter
#' my_chunker_object2 <- chunker(tmp_path_csv, chunksize = 30, sep = ",")
#' # the file can  then be processed as with tab delimiters
#' 
#' # remove temporal file
#' file.remove(tmp_path_csv)
#' 
#' #--------------------------------------------------------#
#' #--- Reading a data frame using column types argument ---#
#' #--------------------------------------------------------#
#' 
#' ## Four types can be passed : "character", "numeric" (aka "double"), "integer", "logical"
#' 
#' # create a 'chunker' object passing the path of the input file.
#' my_chunker_object3 <- chunker(tmp_path, chunksize = 120,
#'  columns_classes = c("numeric", "numeric", "numeric","numeric", "character"))
#' 
#' # read a chunk
#' next_chunk(my_chunker_object3)
#' 
#' # get the chunk
#' get_table(my_chunker_object3)
#' 
#' # read another chunk
#' next_chunk(my_chunker_object3)
#' 
#' # get the number of lines already read
#' get_completed(my_chunker_object3)
#' 
#' 
#' #-------------------------#
#' #--- Reading a matrix  ---#
#' #-------------------------#
#' 
#' my_chunker_object4 <- chunker(tmp_path, chunksize = 30, data_format= "matrix")
#' 
#' # store the chunk as a character matrix in R
#' this_data <- get_table(my_chunker_object4)
#' 
#' 
#' # The package provides a fast generic C++ function for conversion from
#' # matrix (any R type) to data frame
#' this_data_as_df2 <- matrix2df(this_data)
#' 
#' # remove temporal file
#' file.remove(tmp_path)
#' 

#' \dontrun{ 
#' 
#' #----------------------------------#
#' #--- Example with a big table -----#
#' #----------------------------------#
#' 
#' ### Example with a data frame
#' 
#' # create a large data frame, and write it in a temporal directory
#' 
#' tmp_path <- file.path(tempdir(),"big_table.txt")
#' 
#' out <- data.frame(numeric_data = runif(1000000),
#'                   character_data = sample(c("a", "t", "c", "g"), 1000000, 
#'                   replace = TRUE),
#'                   integer_data = sample(1000000),
#'                   bool_data = sample(c(TRUE, FALSE), 1000000, replace = TRUE))
#' 
#' 
#' write.table(out, tmp_path, quote = FALSE)
#' 
#' # create a chunker object, reading in chunks of 10000 lines
#' my_chunker_object5 <- chunker(tmp_path, chunksize = 10000)
#' 
#' next_chunk(my_chunker_object5)
#' data <- get_table(my_chunker_object5) 
#' 
#' # check classes
#' lapply(data,typeof)
#' file.remove(tmp_path)
#' 
#' 
#' ### Example with a matrix
#' 
#' # create a large matrix, and write it in a temporal directory
#' 
#' my_table <- tempfile()
#' write.table(matrix(sample(c("a", "t", "c", "g"), 1000000, replace = TRUE), 
#' 100000, 1000), my_table, quote = FALSE)
#' 
#' # create a chunker object, reading in chunks of 10000 lines
#' my_chunker_object6 <- chunker(my_table, chunksize = 10000, data_format= "matrix")
#'
#' # create a loop to read all the file and make something with it
#' 
#' lines <- 0
#' while(next_chunk(my_chunker_object6))
#' {
#'   data <- get_table(my_chunker_object6) 
#'   
#'   # do something with data, e.g., convert to data frame first
#'   data <- matrix2df(data)
#'   
#'   lines <- lines + nrow(data)
#'   cat("Processed ", lines, "lines\n")
#' }
#' 
#' # remove the temporal file
#' file.remove(my_table)

#' }
#' 
#' @export

setGeneric("chunker", function(path, sep = " ", quoted = FALSE,
                               has_colnames = TRUE, has_rownames = TRUE, 
                               chunksize = 1000L, data_format = c("data.frame", "matrix"),
                               columns_classes = character(0), autodetect = TRUE,
                               scan_rows = 10) {
  
  data_format <- match.arg(data_format)
  
  if(length(columns_classes) != 0) {
    autodetect <- FALSE
  }
  
  if(data_format == "matrix") {
    autodetect <- FALSE
    if(length(columns_classes) > 0) {
      stop("The parameter columns_classes can not be used with output data of type matrix")
    }
  }
  
  if(autodetect) {
    
      if(!has_rownames) {
        x <- read.table(path, nrows = scan_rows, header = has_colnames, 
                        row.names = NULL, stringsAsFactors = FALSE, sep = sep)
      } else {
        x <- read.table(path, nrows = scan_rows, header = has_colnames, 
                        stringsAsFactors = FALSE, sep = sep) 
      }
      # special case with another custom newline
 
    x <- lapply(as.list(x), typeof)
    columns_classes <- unname(unlist(x))
  }

  out <- new("chunker", path, sep, quoted, has_colnames, 
      has_rownames, chunksize, data_format, columns_classes)
  out@attr <- list(path = path, sep = sep, quoted = quoted,
                   has_colnames = has_colnames, has_rownames  = has_rownames,
                   chunksize = chunksize, data_format = data_format,
                   columns_classes = columns_classes, call = match.call())
 out
})
